### PR TITLE
Feature: Optimized uncle jim pdf for the Debasafebags.

### PR DESCRIPTION
**Pull request by f9ert**
image

### CODE REVIEW

**Comment by moneymanolis** [comment]
The displayed character length of 42 in the pdf for the addresses is fine: P2WPKH addresses are always 42 characters long (for more details see here: https://bitcoin.stackexchange.com/questions/100508/can-you-break-down-what-data-is-encoded-into-a-bech32-address). With regtest addresses we are adding two characters ("rt"), thus two characters are missing then, just keep this in mind. Perhaps it makes sense to add this non-trivial background info to the tool-tip belonging to the pdf, @k9ert?

File: src/cryptoadvance/specter/templates/wallet/components/wallet_pdf.jinja [hunk 1]
```diff
@@ -174,47 +174,47 @@
...
             doc.text({{ wallet.name | tojson | safe }}.replace(/(.{55})/g, "$1\n"), 105, currentHeight, {align: 'center'});
             doc.setFontSize(9);
-            doc.text('{{ _("Scan this QR code with Specter-Desktop") }}', 65, currentHeight + 6, {align: 'left'});
-            doc.text("The below addresses are from index 1. ", 65, currentHeight + 12)
-            doc.text("Feel free to use the checkbox as a proof", 65, currentHeight + 18)
-            doc.text("to have verified the address via the Hardwarewallet.", 65, currentHeight + 24)
+            doc.text('{{ _("Scan this QR code with Specter-Desktop") }}', 50, currentHeight + 6, {align: 'left'});
+            doc.text("The below addresses are from index 1. ", 50, currentHeight + 12)
+            doc.text("Feel free to use the checkbox as a proof", 50, currentHeight + 18)
+            doc.text("to have verified the address via the Hardwarewallet.", 50, currentHeight +24)
```
Comment (moneymanolis) [inline src/cryptoadvance/specter/templates/wallet/components/wallet_pdf.jinja:190]:
```suggestion
            doc.text("to have verified the address via a hardware wallet.", 50, currentHeight +24)
```

File: src/cryptoadvance/specter/templates/wallet/components/wallet_pdf.jinja [hunk 2]
```diff
@@ -174,47 +174,47 @@
...
-            doc.addImage(qrCode, 'PNG', 5, currentHeight, 50, 50)
+            doc.addImage(qrCode, 'PNG', 5, currentHeight, 30, 30)
             doc.text({{ wallet.name | tojson | safe }}.replace(/(.{55})/g, "$1\n"), 105, currentHeight, {align: 'center'});
             doc.setFontSize(9);
-            doc.text('{{ _("Scan this QR code with Specter-Desktop") }}', 65, currentHeight + 6, {align: 'left'});
-            doc.text("The below addresses are from index 1. ", 65, currentHeight + 12)
-            doc.text("Feel free to use the checkbox as a proof", 65, currentHeight + 18)
-            doc.text("to have verified the address via the Hardwarewallet.", 65, currentHeight + 24)
+            doc.text('{{ _("Scan this QR code with Specter-Desktop") }}', 50, currentHeight + 6, {align: 'left'});
+            doc.text("The below addresses are from index 1. ", 50, currentHeight + 12)
```
Comment (moneymanolis) [inline src/cryptoadvance/specter/templates/wallet/components/wallet_pdf.jinja:188]:
```suggestion
            doc.text("The addresses below start from index 1. ", 50, currentHeight + 12)
```

**Comment by k9ert** [comment]
> Perhaps it makes sense to add this non-trivial background info to the tool-tip belonging to the pdf, @k9ert?

Great finding! Maybe it also make sense to print it directly on the pdf but just if the network is regtest. Do we also have char-loss for testnet?

**Comment by moneymanolis** [comment]
No, testnet is fine, too, it's just tb:

image

**Comment by moneymanolis** [comment]
I think I would only enable the Uncle Jim PDF feature if the node is set to mainnet, it only poses risks to send to testnet addresses (if the wallet used is not validating correctly ..., see: https://bitcoin.stackexchange.com/questions/18028/what-happens-if-you-send-btc-to-a-testnet-address)

**Comment by k9ert** [comment]
> I think I would only enable the Uncle Jim PDF feature if the node is set to mainnet,

Makes development and testing out stuff a lot more complicated. Don't like the idea so much. I'll try to print it at the bottom of the page...

File: src/cryptoadvance/specter/templates/wallet/components/wallet_pdf.jinja [hunk 3]
```diff
@@ -242,8 +242,14 @@
                 xpubLines = "{{ key.xpub }}".split(/(.{36})/g);
                 xpubLines = xpubLines.filter(line => line !== '');
                 currentHeight += 6
+                doc.setLineDash([2, 2], 0);
+                doc.line(181,0,181,214)
+                doc.line(0,214,181,214)
```
Comment (k9ert) [inline src/cryptoadvance/specter/templates/wallet/components/wallet_pdf.jinja:247]: You're drawing those lines for every key in the wallet. Please move it out of the loop!

**Review by k9ert — APPROVED** [review]

**Pull request merged**